From 48662e55984cfe76ed7656915a799c24ea819105 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:04:13 -0400
Subject: [powerpc] Fix the corrupt r3 error during MCE handling

Message-id: <1410545655-205645-25-git-send-email-dzickus@redhat.com>
Patchwork-id: 94077
O-Subject: [RHEL7 PATCH 024/626] powerpc: Fix the corrupt r3 error during MCE handling.
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit ee1dd1e3dc774cf257012215d996e8e7e370c162
Author: Mahesh Salgaonkar <mahesh@linux.vnet.ibm.com>
Date:   Wed Jul 10 18:32:56 2013 +0530

    powerpc: Fix the corrupt r3 error during MCE handling.

    During Machine Check interrupt on pseries platform, R3 generally points to
    memory region inside RTAS (FWNMI) area. We see r3 corruption because when RTAS
    delivers the machine check exception it passes the address inside FWNMI area
    with the top most bit set. This patch fixes this issue by masking top two bit
    in machine check exception handler.

    Signed-off-by: Mahesh Salgaonkar <mahesh@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/platforms/pseries/ras.c b/arch/powerpc/platforms/pseries/ras.c
index 7b3cbde..721c058 100644
--- a/arch/powerpc/platforms/pseries/ras.c
+++ b/arch/powerpc/platforms/pseries/ras.c
@@ -287,6 +287,9 @@ static struct rtas_error_log *fwnmi_get_errinfo(struct pt_regs *regs)
  unsigned long *savep;
  struct rtas_error_log *h, *errhdr = NULL;
 
+ /* Mask top two bits */
+ regs->gpr[3] &= ~(0x3UL << 62);
+
  if (!VALID_FWNMI_BUFFER(regs->gpr[3])) {
   printk(KERN_ERR "FWNMI: corrupt r3 0x%016lx\n", regs->gpr[3]);
   return NULL;
-- 
1.7.1